Kernel/Devices/Storage/VirtIO/VirtIOBlockDevice.cpp

```diff
@@ -120,7 +120,10 @@ UNMAP_AFTER_INIT ErrorOr<void> VirtIOBlockDevice::initialize_virtio_resources()
     }));
     TRY(setup_queues(1)); // REQUESTQ
     finish_init();
-    return {};
+
+    // This device is not added to DeviceManagement, as we've bypassed try_create_device, so we need to add it ourselves
+    // We can't call the method directly, as it is redeclared as private in the StorageDevice class, thus the cast
+    return static_cast<Kernel::Device*>(this)->after_inserting();
```

Comment:
No, this is happening because we don't follow the usual pattern of construction here.
**This is an horrible hack**, and I'd strongly advise to not go further with this.

Fix this by looking how we construct other `StorageDevice` please. For example, you could do this here:
```c++
UNMAP_AFTER_INIT ErrorOr<NonnullLockRefPtr<VirtIOBlockDevice>> VirtIOBlockDevice::try_create( ... )
{
    auto device = TRY(DeviceManagement::try_create_device<VirtIOBlockDevice>( ... ));
    TRY(device->initialize_virtio_resources());
    return device;
}
```


And then on VirtIOBlockController code, do this:
```c++
    ...
    auto device = TRY(VirtIOBlockDevice::try_create( ... ));
    TRY(device->initialize_virtio_resources());
    ...
```

Reply:
It seems that this happens because the constructor is assumed to be infallible, and here the "constructor" (initialize_virtio_resources) can fail. I guess a better way to fix this would be to:

1. Override after_inserting in VirtIOBlockDevice, move logic from initialize_virtio_resources there, then tailcall to the original method.
2. Make VirtIOBlockController use try_create.

The obvious issue with try_create is that it would insert a partially-initialized device into a global hash table, which is probably not what we want.

Reply:
The constructor should not have anything that can logically fail. Please go forward with option 1, because that's what we do essentially for any other `Device`-derived class in the kernel codebase.

Reply:
Another idea - why don't we make `VirtIOBlockController` class the `VirtIO::Device`? After all, the block device on itself is not really the VirtIO device, right?
That would make it much more reasonable to let the creation of an exposed char device to fail, and the actual handling of VirtIO stuff will move to `VirtIOBlockController` class instead which makes far more sense to me.

Reply:
Because I'd like to make things go further, I wrote a small patch doing what I suggested here:
https://github.com/supercomputer7/serenity/tree/fix-virtio-block-wrong-init-path

Specifically it has only one commit here:
https://github.com/supercomputer7/serenity/commit/22dbd687f462af6220fe47ed9ffbddbe5fb6feff

I think this is the way forward. I also really don't like that we hardcode the max size (i.e. max LBA) of the VirtIO device, but that's a problem for another PR, not this one.